### ISSUE TITLE
Add cache option to avoid processing unchanged files

### ISSUE THREAD

**Issue by mbrookes** [prettier/prettier#5853]
**Environments:**
- Prettier Version: 1.16.4
- Running Prettier via: CLI (package.json script)
- Runtime: Node.js 8.11.3 
- Operating System: MacOS 10.13.6

**Steps to reproduce:**

N/A

**Expected behavior:**
That when provided with an option such as `--cache`, prettier only process new or changed files. See [eslint](https://eslint.org/docs/user-guide/command-line-interface#--cache) for example.

**Actual behavior:**
Prettier processes all files, whether they have changed or not.

**Comment by g-harel** [comment]
I'd like to work on this issue!

I was thinking I can add the check/write about [here](https://github.com/prettier/prettier/blob/5d671f9360eb93911923dd0b93a798bb34cdb7ee/src/cli/util.js#L494)

My plan is to store a hash containing `prettier version + context + options + contents` per file after prettifying the contents. On the next run, if the same hash is produced using the non-formatted contents, it is already pretty.

The data could be stored in memory while running, and written to disk after all files are iterated over. I would like some input on the cache location, I was thinking [this package](https://www.npmjs.com/package/find-cache-dir) looked promising, but it seems [controversial](https://github.com/avajs/find-cache-dir/issues/1). Maybe a `/tmp` file?

**Comment by mbrookes** [comment]
`+ contents`

Or perhaps modification timestamp? That way you don't have to read the file to know if it has changed.

**Comment by lydell** [comment]
@g-harel I'd try out a bunch of different ways to do the caching (in quick and dirty ways) to find which ones are faster and which ones don't make a difference (or even are slower!). Then we can think about the details.

**Comment by j-f1** [comment]
[Here](https://github.com/eslint/eslint/blob/5018378131fd5190bbccca902c0cf4276ee1581a/lib/util/lint-result-cache.js) is ESLint’s cache implementation.

**Comment by kachkaev** [comment]
And [here](https://github.com/gicentre/prettier-plugin-elm/blob/master/src/cache.ts)'s my implementation of cache for [prettier-plugin-elm](https://github.com/gicentre/prettier-plugin-elm), just in case.

**Comment by g-harel** [comment]
Since this change is in the CLI, I thought it would be less obtrusive to add this feature as an extension of `--write`. This also means the formatting result did not need to be cached, only the fact that a file was well formatted _at a given `mtime`_.

[I've opened a draft PR](https://github.com/prettier/prettier/pull/5910) with an implementation and I'd love some feedback!

**Comment by azz** [comment]
Isn't it better to use your source control to determine which files have changed? Otherwise you need to check every single file in the repository.

**Comment by kachkaev** [comment]
@azz repos can use svn or mercurial, so calling git methods won't work for everyone. Besides, git can be unavailable in some CI environments where you want to lint. E.g. in GitLab CI, the repo is restored by the runner and the commands run inside a custom docker image such as `node:alpine`. I does not go with git preinstalled.

**Comment by azz** [comment]
Caching likely won't work in your CI because the files will be freshly cloned at the start of each run, git doesn't track mtimes either.

Personally I always run Prettier through **ESLint** for CI, or [`pretty-quick`](https://www.npmjs.com/package/pretty-quick)/[`lint-staged`](https://www.npmjs.com/package/lint-staged) (there's also [`prettylint`](https://www.npmjs.com/package/prettylint)) locally, so I've never really felt the need for Prettier to have a cache.

**Comment by kachkaev** [comment]
> won't work in your CI

Some CIs allow you to [cache](https://docs.gitlab.com/ee/ci/caching/) selected folders between the build, e.g. `node_modules`. The same can be done with Prettier’s cache to speed-up builds.

I also have Prettier configured to run locally before committing, but still perform `prettier --check ...` inside the CI. This makes sure that unformatted code never slips through, e.g. when someone has temporary disabled git hooks. Bottom line: It'd be great if all Prettier functionality remained orthogonal to source control, e.g. did not depend on calling git commands 🙂

**Comment by azz** [comment]
If you persist the cache file between builds you'll still never get cache hits because all the inode mtimes will have changed between builds, no?

**Comment by kachkaev** [comment]
If the _cache hash_™ depends on mtime and not just file contents – yes, it won't match in CI. Git [does not try](https://git.wiki.kernel.org/index.php/GitFaq#Why_isn.27t_Git_preserving_modification_time_on_files.3F) to sync mtime with commit time etc.

**Comment by mbrookes** [comment]
@g-harel What hash algo are you using? It might be possible to reduce the overhead. (By coincidence I was listening to the Software Engineering Radio podcast in the car today on probabilistic data structures. This naturally included a discussion on non-cryptographic hash algos. If it would help, I can check which ones were mentioned as being the least computationally expensive.)

**Comment by g-harel** [comment]
That's a great podcast!

I'm using `sha1` and encoding as `base64` [following these recommendations](https://medium.com/@chris_72272/what-is-the-fastest-node-js-hashing-algorithm-c15c1a0e164e).

**Comment by mbrookes** [comment]
sha is a cryptographic hash algorithm (designed to be computationally expensive, in order to make brute-forcing hard, so perhaps not the best choice. 

I'm listening to the podcast[1] again... Okay: fnv or murmurhash (or cityhash / farmhash, Google's derivatives of murmurhash). All have js libs / bindings, but fnv is apparently very simple to implement, so you wouldn't even need an external lib.

Interested to hear how much difference it makes!

[1] http://www.se-radio.net/2019/02/se-radio-episode-358-probabilistic-data-structure-for-big-data-problems/ @ ~10:00

**Comment by g-harel** [comment]
I was trying to avoid adding a dependency, but trying those hash implementations, I don't see a significant change in the time it takes.

I also [profiled a "warm" run](https://gist.github.com/g-harel/f0c700621e9afc20f8a26c7cdef2d421) and it looks like the hashing is about 0.5% of CPU time.

**Comment by mbrookes** [comment]
Huh, interesting. Well thanks for indulging me! I really appreciate you picking up this feature request. The important thing is that it reduces time for repetitive passes over the same data, so eliminating this option is a bonus.

**Comment by g-harel** [comment]
I've marked [the PR](#5910 ) as ready!

**Comment by duailibe** [comment]
@kachkaev 

> repos can use svn or mercurial, so calling git methods won't work for everyone

I believe what @azz suggested was adding this behavior in CI config, not in Prettier itself. For example, I know danger.js has specific APIs to retrieve which files were changed in a PR so you could call Prettier for those specific files.

In that case, it wouldn't be necessary to add this in Prettier

**Comment by mbrookes** [comment]
Depending on CI doesn’t help with cli usage though...

**Comment by evilebottnawi** [comment]
Here some notices about latest implementation https://github.com/prettier/prettier/issues/6577#issue-500554660

**Comment by nojvek** [comment]
Would love to say —cache would be nice. I would caution against using any git commands to do heave lifting. 

In a lot of CI environments on partial things get checked out. There’s just a bunch of files. No git. 

For that the fastest way is to quickly compare {relativeFilePath: hash(options + version + content)} and have the map saved in a file like .prettiercache as json. 

There’s a very fast hash function called xxHash. It’s not meant to be cryptographic but to compare file/content integrity. So it’s a perfect fit. Runs at ram speed too. https://github.com/Cyan4973/xxHash

This would give a simple yet fast implementation. You could additionally have an optimization to store two values. Hash(version + options + file size + file last mod). comparing file sizes and last modified since fs.stat is much faster than reading files. 

But it has an edge case of sometimes getting the wrong output if someone manually modified last mod.

**Comment by mbrookes** [comment]
@evilebottnawi I think at one point you said you were going to work on cleaning up the reverted PR for release in 1.20, then possibly to work with @g-harel if he resubmitted his PR, but it looks like neither happened in the end. Is there any plan to revisit this? If not, I'll close the issue. Thanks!

**Comment by evilebottnawi** [comment]
@mbrookes it is still in our plans, no need to close issue, you can send a PR :smile:

**Comment by mbrookes** [comment]
> you can send a PR

Fair comment 😆  but I don't think I'd be able to do it justice, and would end up taking more of the maintainers time than leaving it to someone competent!

**Comment by chdsbd** [comment]
Would we need to change the Prettier API to support this feature?

With the [current API](https://github.com/prettier/prettier/blob/554b15473dd4032a036d7db91a8f579e624c9822/docs/api.md#prettierformatsource--options) users provide text. But I think if we want to cache within Prettier, we'd need to change the API to accept file paths.

**Comment by RomanLubushkin** [comment]
@mbrookes @g-harel 
Hey guys, is there any chance to see this amazing improvement live?
Without cache, Prettier becomes a slowing part of a dev process.
Specifically for my project:
```
Prettier - Done in 46.65s.
ESlint (with cache enabled) - Done in 1.80s.
```

**Comment by Faolain** [comment]
Any updates here?

**Comment by andersk** [comment]
The API already has `format(text, { filepath })`. Although `filepath` is optional and presumably relative to the current directory, that should be good enough to speed up normal usage with a cache keyed on `path.resolve(filepath)` when it exists.

**Comment by SalvatorePreviti** [comment]
No updates here?

**Comment by ronjouch** [comment]
Of potential interest to commenters/subscribers of this issue: [Speeding up Prettier locally and on your CI with dprint](https://david.deno.dev/posts/faster-prettier-with-dprint/), and [HN post](https://news.ycombinator.com/item?id=31160722).

---

**TL;DR**: dprint, a prettier-like of the Deno world, has a cache option. Its formatting rules are not behavior-identical to prettier, but they are swappable thanks to a prettier plugin. So, with its cache option, results are:

- 1st run: 40s (prettier) → 35s (dprint + prettier plugin + cold cache)
- subsequent runs: 40s (prettier) → 1s (dprint + prettier plugin + warm cache)

But there's nothing fundamentally better in dprint, what matters is the cache. Would be great if it was built-in prettier (what this issue is all about).

**Comment by mbrookes** [comment]
Hard to believe it's been over three years since I opened this issue yet prettier, despite its initial promise, still hasn't implemented this.

@ronjouch I saw the HN story but hadn't made the connection, so thanks. 39s might not sound like much, but for a large repo with frequent CI builds it adds up considerably. Will report back.

**Comment by alexander-akait** [comment]
You can help and send a PR, it's not blocked by something

**Comment by mbrookes** [comment]
https://github.com/prettier/prettier/issues/5853#issuecomment-686065572

And, I saw what happened the first time around... 😁
